src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpRequestCallback.cs

```diff
@@ -0,0 +1,360 @@
+                    ref certHandleSize))
+                {
+                    int lastError = Marshal.GetLastWin32Error();
+                    throw WinHttpException.CreateExceptionUsingError(lastError);
+                }
+                
+                // Create a managed wrapper around the certificate handle. Since this results in duplicating
+                // the handle, we will close the original handle after creating the wrapper.
+                var serverCertificate = new X509Certificate2(certHandle);
+                Interop.Crypt32.CertFreeCertificateContext(certHandle);
+
+                X509Chain chain = null;
+                SslPolicyErrors sslPolicyErrors;
+
+                try
+                {
+                    WinHttpCertificateHelper.BuildChain(
+                        serverCertificate,
+                        state.RequestMessage.RequestUri.Host,
+                        state.CheckCertificateRevocationList,
+                        out chain,
+                        out sslPolicyErrors);
+
+                    bool result = state.ServerCertificateValidationCallback(
+                        state.RequestMessage,
+                        serverCertificate,
+                        chain,
+                        sslPolicyErrors);
+                    if (!result)
+                    {
+                        throw WinHttpException.CreateExceptionUsingError(
+                            (int)Interop.WinHttp.ERROR_WINHTTP_SECURE_FAILURE);
+                    }
+                }
+                finally
+                {
+                    if (chain != null)
+                    {
+                        chain.Dispose();
+                    }
+                }
+            }
+        }
+
+        private static void OnRequestError(WinHttpRequestState state, Interop.WinHttp.WINHTTP_ASYNC_RESULT asyncResult)
+        {
+            WinHttpTraceHelper.TraceAsyncError("OnRequestError", asyncResult);
+            
+            Debug.Assert(state != null, "OnRequestError: state is null");
+
+            var innerException = WinHttpException.CreateExceptionUsingError((int)asyncResult.dwError);
+
+            switch ((uint)asyncResult.dwResult.ToInt32())
+            {
+                case Interop.WinHttp.API_SEND_REQUEST:
+                    state.TcsSendRequest.TrySetException(innerException);
+                    break;
+                    
+                case Interop.WinHttp.API_RECEIVE_RESPONSE:
+                    if (asyncResult.dwError == Interop.WinHttp.ERROR_WINHTTP_RESEND_REQUEST)
+                    {
+                        state.RetryRequest = true;
+                        state.TcsReceiveResponseHeaders.TrySetResult(false);
+                    }
+                    else if (asyncResult.dwError == Interop.WinHttp.ERROR_WINHTTP_CLIENT_AUTH_CERT_NEEDED)
+                    {
+                        // WinHttp will automatically drop any client SSL certificates that we
```

Comment:
There 3 states in WinHTTP for client certificates and how WinHTTP will satisfy the demand/request from the server: 

1. I don't have a cert.  You need to tell me what to do.
2. I have a real cert that I can send to the server.
3. There is no cert to send to the server.  I will send a NULL cert to the server.

The comment:

>WinHttp will automatically drop any client SSL certificates

means we go back to state 1.

For redirects, we don't want to send a real cert (if we had one), we move to state 3 by setting a NULL cert.

Setting a NULL cert is NOT the same thing as state 1.